README.md

```diff
@@ -425,6 +430,78 @@ def xor(*args: typing.Union[bytes, typing.Iterable[bytes]]) -> bytes:
     )
 ```
 
+## Basic Math
+
+### Modular Arithmetic
+
+- https://en.wikipedia.org/wiki/Modular_arithmetic
+- https://docs.python.org/3.10/reference/datamodel.html#object.__mod__
+- https://docs.python.org/3.10/library/operator.html#operator.mod
+
+In the example below `9 mod 5` is congruent to `4`, because `4` is the
+remainder when `9` is divided by `5`.
+
+```python
+>>> 9 % 5
+4
+>>> [i % 5 for i in range(10)]
+[0, 1, 2, 3, 4, 0, 1, 2, 3, 4]
+```
+
+### Exponentiation
+
+- https://docs.python.org/3.10/reference/datamodel.html#object.__pow__
+- https://docs.python.org/3.10/library/operator.html#operator.pow
+
+In some cases, you [need to be careful how you do
+exponentiation](https://stackoverflow.com/a/18965845), but for most purposes,
+the standard `**` operator is adequate (particularly when you don't have to
+worry about overflow).
+
+```python
+>>> 2 ** 3
+8
+>>> 16 ** .5
+4.0
+```
+
+You can also use the built-in `pow` function:
+
+```python
+>>> pow(2, 3)
+8
+>>> pow(16, .5)
+4.0
+```
+
+### Modular Exponentiation
+
+- https://docs.python.org/3/library/functions.html#pow
+
+While it is possible to perform modular exponentiation using the `**` operator,
+it is not recommended unless working with small integers:
+
+```python
+>>> 2 ** 8 % 3
+1
+```
+
+The above expression will first perform the full exponentiation before
+performing any modulation, which can be very expensive for large exponents. The
+built-in `pow` function, on the other hand, takes in a modulus as an optional
+third argument, and performs modulation on each round of exponentiation,
+keeping the internal working values bounded by the modulus. This makes the
+built-in `pow` function _much_ more efficient when working with large
+exponents:
```

Comment:
its actually fun exercise to implement pow yourself 😄

Reply:
yeah! bruce schneier has a really nice algorithm that's only logarithmic in the size of the exponent!

https://github.com/cs-gy6903/2022-fall-grading/blob/main/rust-solutions/src/lib.rs#L71-L87